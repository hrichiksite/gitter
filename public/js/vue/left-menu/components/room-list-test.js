const Vue = require('vue');
const mount = require('../../__test__/vuex-mount');

const { default: RoomList } = require('./room-list.vue');

const { createSerializedRoomFixture } = require('../../__test__/fixture-helpers');

describe('room-list', () => {
  it('matches snapshot', () => {
    const { wrapper } = mount(RoomList, { rooms: [] });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('matches snapshot when mobile', () => {
    const { wrapper } = mount(RoomList, { rooms: [] }, store => {
      store.state.isMobile = true;
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('matches snapshot with some rooms', () => {
    const activeRoom = createSerializedRoomFixture('community/room2');
    const { wrapper } = mount(
      RoomList,
      {
        rooms: [
          createSerializedRoomFixture('community/room1'),
          activeRoom,
          createSerializedRoomFixture('community/room3')
        ]
      },
      store => {
        store.state.displayedRoomId = activeRoom.id;
      }
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('matches snapshot with some favourite rooms', () => {
    const { wrapper } = mount(RoomList, {
      rooms: [
        createSerializedRoomFixture('community/room1'),
        createSerializedRoomFixture('community/room2'),
        {
          ...createSerializedRoomFixture('community/favourite-room1'),
          favourite: 1
        },
        createSerializedRoomFixture('community/room3'),
        {
          ...createSerializedRoomFixture('community/favourite-room2'),
          favourite: 2
        }
      ]
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('calls store action "updatefavouriteDraggingInProgress" -> true when dragging starts', async () => {
    const { wrapper, stubbedActions } = mount(RoomList, {
      rooms: []
    });

    wrapper.vm.dragStart();

    expect(stubbedActions.updatefavouriteDraggingInProgress).toHaveBeenCalledWith(
      expect.anything(),
      true,
      undefined
    );
  });

  it('calls store action "updatefavouriteDraggingInProgress" -> false when dragging starts', async () => {
    const { wrapper, stubbedActions } = mount(RoomList, { rooms: [] });

    wrapper.vm.dragEnd();

    expect(stubbedActions.updatefavouriteDraggingInProgress).toHaveBeenCalledWith(
      expect.anything(),
      false,
      undefined
    );
  });

  describe('moving a favourite room', () => {
    const favouriteRoom1 = {
      ...createSerializedRoomFixture('community/favourite-room1'),
      favourite: 1
    };
    const favouriteRoom2 = {
      ...createSerializedRoomFixture('community/favourite-room2'),
      favourite: 2
    };
    const favouriteRoom3 = {
      ...createSerializedRoomFixture('community/favourite-room3'),
      favourite: 3
    };
    const favouriteRoom4 = {
      ...createSerializedRoomFixture('community/favourite-room4'),
      favourite: 4
    };

    let wrapper, stubbedActions, store;
    beforeEach(() => {
      const mountResult = mount(
        RoomList,
        {
          rooms: [favouriteRoom1, favouriteRoom2, favouriteRoom3, favouriteRoom4]
        },
        store => {
          store.state.roomMap = {
            [favouriteRoom1.id]: favouriteRoom1,
            [favouriteRoom2.id]: favouriteRoom2,
            [favouriteRoom3.id]: favouriteRoom3,
            [favouriteRoom4.id]: favouriteRoom4
          };
        }
      );
      wrapper = mountResult.wrapper;
      stubbedActions = mountResult.stubbedActions;
      store = mountResult.store;
    });

    it('calls store action "updateRoomFavourite" with (maxfavourite + 1) when favourite room moved from first to last', async () => {
      const oldIndex = 0;
      const newIndex = wrapper.vm.favouriteRooms.length - 1;
      wrapper.vm.favouriteMoved({ oldIndex, newIndex });

      expect(stubbedActions.updateRoomFavourite).toHaveBeenCalledWith(
        expect.anything(),
        {
          id: wrapper.vm.favouriteRooms[oldIndex].id,
          favourite: wrapper.vm.favouriteRooms[newIndex].favourite + 1
        },
        undefined
      );

      // Add a bit of integration test just to make sure things are flowing all the way to the store correctly
      await Vue.nextTick();
      expect(
        Object.values(store.state.roomMap).map(({ id, favourite }) => ({ id, favourite }))
      ).toMatchSnapshot();
    });

    it('calls store action "updateRoomFavourite" with favourite of the sibling it bumps when favourite room moved from last to first', async () => {
      const oldIndex = wrapper.vm.favouriteRooms.length - 1;
      const newIndex = 0;
      wrapper.vm.favouriteMoved({ oldIndex, newIndex });

      expect(stubbedActions.updateRoomFavourite).toHaveBeenCalledWith(
        expect.anything(),
        {
          id: wrapper.vm.favouriteRooms[oldIndex].id,
          favourite: wrapper.vm.favouriteRooms[newIndex].favourite
        },
        undefined
      );

      // Add a bit of integration test just to make sure things are flowing all the way to the store correctly
      await Vue.nextTick();
      expect(
        Object.values(store.state.roomMap).map(({ id, favourite }) => ({ id, favourite }))
      ).toMatchSnapshot();
    });

    it('calls store action "updateRoomFavourite" with favourite of the sibling it bumps when favourite room moved upwards (in the middle)', async () => {
      const oldIndex = wrapper.vm.favouriteRooms.length - 1;
      const newIndex = 1;
      wrapper.vm.favouriteMoved({ oldIndex, newIndex });

      expect(stubbedActions.updateRoomFavourite).toHaveBeenCalledWith(
        expect.anything(),
        {
          id: wrapper.vm.favouriteRooms[oldIndex].id,
          favourite: wrapper.vm.favouriteRooms[newIndex].favourite
        },
        undefined
      );

      // Add a bit of integration test just to make sure things are flowing all the way to the store correctly
      await Vue.nextTick();
      expect(
        Object.values(store.state.roomMap).map(({ id, favourite }) => ({ id, favourite }))
      ).toMatchSnapshot();
    });

    it('calls store action "updateRoomFavourite" with favourite of sibling it bumps accountnig for itself when favourite room moved downwards (in the middle)', async () => {
      const oldIndex = 0;
      const newIndex = 2;
      wrapper.vm.favouriteMoved({ oldIndex, newIndex });

      expect(stubbedActions.updateRoomFavourite).toHaveBeenCalledWith(
        expect.anything(),
        {
          id: wrapper.vm.favouriteRooms[oldIndex].id,
          favourite: wrapper.vm.favouriteRooms[newIndex + 1].favourite
        },
        undefined
      );

      // Add a bit of integration test just to make sure things are flowing all the way to the store correctly
      await Vue.nextTick();
      expect(
        Object.values(store.state.roomMap).map(({ id, favourite }) => ({ id, favourite }))
      ).toMatchSnapshot();
    });
  });

  describe('adding a favourite room (dragging from normal to favourite list)', () => {
    const favouriteRoom1 = {
      ...createSerializedRoomFixture('community/favourite-room1'),
      favourite: 1
    };
    const favouriteRoom2 = {
      ...createSerializedRoomFixture('community/favourite-room2'),
      favourite: 2
    };
    const favouriteRoom3 = {
      ...createSerializedRoomFixture('community/favourite-room3'),
      favourite: 3
    };
    const room1 = createSerializedRoomFixture('community/room1');

    it('calls store action "updateRoomFavourite" with new room favourite', async () => {
      const { wrapper, stubbedActions, store } = mount(
        RoomList,
        {
          rooms: [favouriteRoom1, favouriteRoom2, favouriteRoom3, room1]
        },
        store => {
          store.state.roomMap = {
            [favouriteRoom1.id]: favouriteRoom1,
            [favouriteRoom2.id]: favouriteRoom2,
            [favouriteRoom3.id]: favouriteRoom3,
            [room1.id]: room1
          };
        }
      );
      const oldIndex = 0;
      const newIndex = 2;
      wrapper.vm.favouriteAdded({ oldIndex, newIndex });

      expect(stubbedActions.updateRoomFavourite).toHaveBeenCalledWith(
        expect.anything(),
        {
          id: wrapper.vm.normalRooms[oldIndex].id,
          favourite: wrapper.vm.favouriteRooms[newIndex].favourite
        },
        undefined
      );

      // Add a bit of integration test just to make sure things are flowing all the way to the store correctly
      await Vue.nextTick();
      expect(
        Object.values(store.state.roomMap).map(({ id, favourite }) => ({ id, favourite }))
      ).toMatchSnapshot();
    });
  });

  describe('removing a favourite room (dragging from favourite list to normal list)', () => {
    const favouriteRoom1 = {
      ...createSerializedRoomFixture('community/favourite-room1'),
      favourite: 1
    };
    const room1 = createSerializedRoomFixture('community/room1');
    const room2 = createSerializedRoomFixture('community/room2');
    const room3 = createSerializedRoomFixture('community/room3');

    it('calls store action "updateRoomFavourite" with removed room favourite', async () => {
      const { wrapper, stubbedActions, store } = mount(
        RoomList,
        {
          rooms: [favouriteRoom1, room1, room2, room3]
        },
        store => {
          store.state.roomMap = {
            [favouriteRoom1.id]: favouriteRoom1,
            [room1.id]: room1,
            [room2.id]: room2,
            [room3.id]: room3
          };
        }
      );
      const oldIndex = 0;
      const newIndex = 1;
      wrapper.vm.favouriteRemoved({ oldIndex, newIndex });

      expect(stubbedActions.updateRoomFavourite).toHaveBeenCalledWith(
        expect.anything(),
        {
          id: wrapper.vm.favouriteRooms[oldIndex].id,
          favourite: false
        },
        undefined
      );

      // Add a bit of integration test just to make sure things are flowing all the way to the store correctly
      await Vue.nextTick();
      expect(
        Object.values(store.state.roomMap).map(({ id, favourite }) => ({ id, favourite }))
      ).toMatchSnapshot();
    });
  });
});
