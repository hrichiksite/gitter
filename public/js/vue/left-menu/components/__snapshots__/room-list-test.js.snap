// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`room-list adding a favourite room (dragging from normal to favourite list) calls store action "updateRoomFavourite" with new room favourite 1`] = `
Array [
  Object {
    "favourite": 1,
    "id": "5a8739841543b98772a686a9-community/favourite-room1",
  },
  Object {
    "favourite": 2,
    "id": "5a8739841543b98772a686a9-community/favourite-room2",
  },
  Object {
    "favourite": 4,
    "id": "5a8739841543b98772a686a9-community/favourite-room3",
  },
  Object {
    "favourite": 3,
    "id": "5a8739841543b98772a686a9-community/room1",
  },
]
`;

exports[`room-list matches snapshot 1`] = `
<div>
  <draggable-stub
    class="favourite-room-list"
    clone="function(t){return t}"
    element="div"
    group="rooms"
    tag="ul"
  />
   
  <draggable-stub
    class="room-list"
    clone="function(t){return t}"
    element="div"
    group="rooms"
    tag="ul"
  />
</div>
`;

exports[`room-list matches snapshot when mobile 1`] = `
<div>
  <draggable-stub
    class="favourite-room-list"
    clone="function(t){return t}"
    disabled="true"
    element="div"
    group="rooms"
    tag="ul"
  />
   
  <draggable-stub
    class="room-list"
    clone="function(t){return t}"
    disabled="true"
    element="div"
    group="rooms"
    tag="ul"
  />
</div>
`;

exports[`room-list matches snapshot with some favourite rooms 1`] = `
<div>
  <draggable-stub
    class="favourite-room-list"
    clone="function(t){return t}"
    element="div"
    group="rooms"
    tag="ul"
  >
    <list-item-stub
      item="[object Object]"
      type="all"
    />
    <list-item-stub
      item="[object Object]"
      type="all"
    />
  </draggable-stub>
   
  <draggable-stub
    class="room-list"
    clone="function(t){return t}"
    element="div"
    group="rooms"
    tag="ul"
  >
    <list-item-stub
      item="[object Object]"
      type="all"
    />
    <list-item-stub
      item="[object Object]"
      type="all"
    />
    <list-item-stub
      item="[object Object]"
      type="all"
    />
  </draggable-stub>
</div>
`;

exports[`room-list matches snapshot with some rooms 1`] = `
<div>
  <draggable-stub
    class="favourite-room-list"
    clone="function(t){return t}"
    element="div"
    group="rooms"
    tag="ul"
  />
   
  <draggable-stub
    class="room-list"
    clone="function(t){return t}"
    element="div"
    group="rooms"
    tag="ul"
  >
    <list-item-stub
      item="[object Object]"
      type="all"
    />
    <list-item-stub
      active="true"
      item="[object Object]"
      type="all"
    />
    <list-item-stub
      item="[object Object]"
      type="all"
    />
  </draggable-stub>
</div>
`;

exports[`room-list moving a favourite room calls store action "updateRoomFavourite" with (maxfavourite + 1) when favourite room moved from first to last 1`] = `
Array [
  Object {
    "favourite": 5,
    "id": "5a8739841543b98772a686a9-community/favourite-room1",
  },
  Object {
    "favourite": 2,
    "id": "5a8739841543b98772a686a9-community/favourite-room2",
  },
  Object {
    "favourite": 3,
    "id": "5a8739841543b98772a686a9-community/favourite-room3",
  },
  Object {
    "favourite": 4,
    "id": "5a8739841543b98772a686a9-community/favourite-room4",
  },
]
`;

exports[`room-list moving a favourite room calls store action "updateRoomFavourite" with favourite of sibling it bumps accountnig for itself when favourite room moved downwards (in the middle) 1`] = `
Array [
  Object {
    "favourite": 4,
    "id": "5a8739841543b98772a686a9-community/favourite-room1",
  },
  Object {
    "favourite": 2,
    "id": "5a8739841543b98772a686a9-community/favourite-room2",
  },
  Object {
    "favourite": 3,
    "id": "5a8739841543b98772a686a9-community/favourite-room3",
  },
  Object {
    "favourite": 5,
    "id": "5a8739841543b98772a686a9-community/favourite-room4",
  },
]
`;

exports[`room-list moving a favourite room calls store action "updateRoomFavourite" with favourite of the sibling it bumps when favourite room moved from last to first 1`] = `
Array [
  Object {
    "favourite": 2,
    "id": "5a8739841543b98772a686a9-community/favourite-room1",
  },
  Object {
    "favourite": 3,
    "id": "5a8739841543b98772a686a9-community/favourite-room2",
  },
  Object {
    "favourite": 4,
    "id": "5a8739841543b98772a686a9-community/favourite-room3",
  },
  Object {
    "favourite": 1,
    "id": "5a8739841543b98772a686a9-community/favourite-room4",
  },
]
`;

exports[`room-list moving a favourite room calls store action "updateRoomFavourite" with favourite of the sibling it bumps when favourite room moved upwards (in the middle) 1`] = `
Array [
  Object {
    "favourite": 1,
    "id": "5a8739841543b98772a686a9-community/favourite-room1",
  },
  Object {
    "favourite": 3,
    "id": "5a8739841543b98772a686a9-community/favourite-room2",
  },
  Object {
    "favourite": 4,
    "id": "5a8739841543b98772a686a9-community/favourite-room3",
  },
  Object {
    "favourite": 2,
    "id": "5a8739841543b98772a686a9-community/favourite-room4",
  },
]
`;

exports[`room-list removing a favourite room (dragging from favourite list to normal list) calls store action "updateRoomFavourite" with removed room favourite 1`] = `
Array [
  Object {
    "favourite": false,
    "id": "5a8739841543b98772a686a9-community/favourite-room1",
  },
  Object {
    "favourite": undefined,
    "id": "5a8739841543b98772a686a9-community/room1",
  },
  Object {
    "favourite": undefined,
    "id": "5a8739841543b98772a686a9-community/room2",
  },
  Object {
    "favourite": undefined,
    "id": "5a8739841543b98772a686a9-community/room3",
  },
]
`;
